numpyro/distributions/distribution.py

```diff
@@ -106,6 +106,21 @@ class Distribution(object):
     reparametrized_params = []
     _validate_args = False
 
+    # register Distribution as a pytree
+    # ref: https://github.com/google/jax/issues/2916
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        tree_util.register_pytree_node(cls,
+                                       cls.tree_flatten,
+                                       cls.tree_unflatten)
+
+    def tree_flatten(self):
```

Comment:
This default implementation works because most distributions only need ndarray args at their constructor (some special cases such as `event_dim` in Delta or `covariance_matrix=None` in MVN or `dimension=4` in LKJ,...). I will try to use `inspect` module here to guarantee that the order of args is matched.

Reply:
Actually, another possibility is to sort by key here:

```python
    return tuple(getattr(self, param) for param in sorted(self.arg_constraints.keys())), None
```

and then, for unflatten:
```python
    return cls(**dict(zip(sorted(self.arg_constraints.keys()), params)))
```

Would that work?

Reply:
Oh, nice trick at the unflatten! Yes, that should work. Many thanks!